### ISSUE TITLE
Sik-Ho Tang | Review: STN -- Spatial Transformer Network (Image Classification).

### ISSUE THREAD

**Issue by NorbertZheng** [NorbertZheng/read-papers#93]
Sik-Ho Tang. [Review: STN — Spatial Transformer Network (Image Classification)](https://towardsdatascience.com/review-stn-spatial-transformer-network-image-classification-d3cbd98a70aa).

**Comment by NorbertZheng** [comment]
## Overview
In this story, Spatial Transformer Network (STN), by Google DeepMind, is briefly reviewed. STN helps to crop out and scale-normalizes the appropriate region, which can simplify the subsequent classification task and lead to better classification performance as below:
![1_P_nv_a_Q3LqM9d10XGE3TQ](https://github.com/NorbertZheng/read-papers/assets/37235023/b2da75a9-e404-47dc-ae19-01e3e7f163a5)

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/e44be2c3-6551-409d-8a91-4f06a426fea4)
(a) Input Image with Random Translation, Scale, Rotation, and Clutter, (b) STN Applied to Input Image, (c) Output of STN, (d) Classification Prediction.

It is published in 2015 NIPS with more than 1300 citations. **Spatial transformation such as affine transformation and homography registration has been studied for decades.**
 - But in this paper, spatial transformation is coped with neural network.

With learning-based spatial transformation, transformation is applied conditioned on input or feature map. And it is highly related to another paper called “[Deformable Convolutional Networks](https://arxiv.org/abs/1703.06211)” (2017 ICCV). Thus, I decided to read this first.

**Comment by NorbertZheng** [comment]
## Quick Review on Spatial Transformation Matrices
There are mainly 3 transformation learnt by STN in the paper. Indeed, more sophisticated transformation can also be applied as well.
### Affine Transformation
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/7c9c31f9-ebda-436e-a496-f4db2b1a6b1d)
Affine Transform.

Depending the values in the matrix, we can transform (X1, Y1) to (X2, Y2) with different effects, as follows:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/5a30489b-affb-4af7-9021-0c7b867e4712)
Translation, Scaling, Rotation, and Shearing.

If interested, please Google “Registration”, “Homography Matrix”, or “Affine Transform”.

**Comment by NorbertZheng** [comment]
### Projective Transformation
Projective transformation can also be learnt in STN as below:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/aff89e57-a97e-467e-be39-89976bc4d630)
Projective Transformation.

**Comment by NorbertZheng** [comment]
### Thin Plate Spline (TPS) Transformation
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/6ad8d3f3-7ae6-4721-873c-70adce3ba1b4)
Thin Plate Spline (TPS) Transformation.

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/11b44ac0-aea5-44ab-8294-9b97f59f24f9)
An example.

For TPS transformation, it is more complicated compared with the previous two transformation. (I have learnt affine and projective mapping before, but I haven’t touched about TPS, if there is mistakes, please tell me.)

To be brief, suppose we have a point (x, y) at a location other than the input points (xi, yi), we use the equations at the right to transform the point based on a bias, weighted sum of x and y, and a function of distance between (x, y) and (xi, yi). (Here, a radial basis function RBF.)

Therefore, if we use TPS, the network needs to learn a0, a1, a2, b0, b1, b2, Fi, and Gi, which are 6+2N number of parameters.

 - As we can see, a more flexible or higher degree of freedom of deformation or transformation can be achieved by TPS.

**Comment by NorbertZheng** [comment]
## Spatial Transformer Network (STN)
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/100f5074-7116-42ee-8272-5c1f6ec73948)
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/0b76874e-bba9-4db4-b541-36564c7ebb09)
Affine Transformation.

STN is composed of Localisation Net, Grid Generator and Sampler.
### Localisation Net
 - **With input feature map $U$, with width $W$, height $H$ and $C$ channels, outputs are $\theta$, the parameters of transformation $T_{\theta}$.**

It can be learnt as affine transform as above. Or to be more constrained such as the used for attention which only contains scaling and translation as below:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/b8e84541-11d0-451a-87e8-0748d8e2c137)
Only scaling and translation.

### Grid Generator
 - Suppose we have a regular grid $G$, this $G$ is a set of points with target coordinates ($x^t_i$, $y^t_i$).

 - Then we apply transformation $T_{\theta}$ on $G$, i.e. $T_{\theta}(G)$.

 - After $T_{\theta}(G)$, a set of points with destination coordinates ($x^t_i$, $y^t_i$) is outputted. These points have been altered based on the transformation parameters. It can be Translation, Scale, Rotation or More Generic Warping depending on how we set $\theta$ as mentioned above.

### Sampler
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/3492352f-e408-49ad-88d3-7c879da2e899)
(a) Identity Transformation, (b) Affine Transformation.

 - Based on the new set of coordinates ($x^t_i$, $y^t_i$), we generate a transformed output feature map $V$. This $V$ is translated, scaled, rotated, warped, projective transformed or affined, whatever.
 - It is noted that STN can be applied to not only input image, but also intermediate feature maps.

**Comment by NorbertZheng** [comment]
Subject Layer in MEG decoding is just an affine transformation. We don't have expert domain knowledge, maybe affine transformation is already enough.

**Comment by NorbertZheng** [comment]
## Sampling Kernel
As we can see in the example above, if we need to sample a transformed grid, we got sampling problem, how we sampling those sub-pel positions are depending on what sampling kernel we about to use.

 - General Form:
   ![image](https://github.com/NorbertZheng/read-papers/assets/37235023/858215f2-8a3a-465a-a49f-a652bd3da70f)
 - Integer Sampling Kernel (by rounding to the nearest integer):
   ![image](https://github.com/NorbertZheng/read-papers/assets/37235023/132128e7-e85f-4cef-844c-5cb3aab4d18b)
 - Bilinear Sampling Kernel:
   ![image](https://github.com/NorbertZheng/read-papers/assets/37235023/d008965c-4bba-4819-b184-d9ec50450e1c)
   It is a (sub-)differentiable sampling mechanism so that it is convenient for backpropagation:
   ![image](https://github.com/NorbertZheng/read-papers/assets/37235023/8ecdc93b-3d89-4047-8e3e-7d035fcd479e)

**Comment by NorbertZheng** [comment]
## Experimental Results
### Distorted MNIST
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/1f2a64d4-b389-400a-a3eb-ab9bef1c8ff1)
Errors of distorted MNIST datasets (Left), Some examples that are failed in CNN but successfully classified in STN (Right).

 - Distortion applied: TC: translated and cluttered, R: rotated, RTS: rotated, translated, and scaled, P: projective distortion, E: elastic distortion.
 - Spatial transformers: Aff: Affine Transformation, Proj: Projective Transformation, TPS: Thin Plate Spline Transformation.
 - FCN: FCN here means Fully Connected Network without convolutions (It is NOT Fully Convolutional Network here.)

As we can see, ST-FCN outperforms FCN and ST-CNN outperforms CNN. And ST-CNN is consistently better than ST-FCN in all settings.

**Comment by NorbertZheng** [comment]
### SVHN (Street View House Number)
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/a4d4c02e-5878-4f10-ba3d-0648c37b7153)
Errors of SVHN datasets (Left), Some examples use in ST-CNN (Right).

 - ST-CNN Single: Only one ST at the beginning of network.
 - ST-CNN Multi: one ST before each conv.
 - Affine transformation is used here.

Similarly, ST-CNN outperforms Maxout and CNN. (I have a very brief introduction of Maxout in [NoC](https://medium.com/datadriveninvestor/review-noc-winner-in-2015-coco-ilsvrc-detection-object-detection-d5cc84e372a), please read it if interested.) And ST-CNN Multi outperforms ST-CNN Single a bit.

**Comment by NorbertZheng** [comment]
### Fine-Grained Classification
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/67b81106-f79b-41b5-875f-0ac281985883)
Fine-Grained Bird Classification. Accuracy(left), 2×ST-CNN (Top Right Row), 4×ST-CNN (Bottom Right Row).

 - Here, ImageNet Pre-trained [Inception-v2](https://medium.com/@sh.tsang/review-batch-normalization-inception-v2-bn-inception-the-2nd-to-surpass-human-level-18e2d0f56651) is used as backbone for classifying 200 species, which has 82.3% accuracy.
 - 2/4×ST-CNN: 2/4 parallel STs, with higher accuracy.

It is interesting that one ST (red) has learnt to be a head detector, with other 3 STs (green) learn the central part of the body of a bird.

**Comment by NorbertZheng** [comment]
## Some Other Tasks
### MNIST Addition
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/1d583e5a-054e-4198-a7d2-f6abc5bbfe83)
MNIST Addition.

2×ST-CNN: It is interesting that each of ST learns to transform each of a digit though each ST also receives two input digits.

**Comment by NorbertZheng** [comment]
### Co-localisation
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/47303220-db85-4fda-bbe7-d1cac2d5590c)
Co-localisation.

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/1feec055-2dbb-4cf5-b4e0-a595f252eb12)
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/c563b395-06ec-4e8e-bedc-c414200f6924)
Triplet loss: Hinge loss is used to enforce the distance between the two outputs of the ST to be less than the distance to a random crop, hoping to encourage the spatial transformer to localise the common objects.

**Comment by NorbertZheng** [comment]
### Higher Dimensional Transformers
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/af7ae94d-ef12-4c1c-ab44-6bff7ab34177)
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/1ab17eed-51ec-4a60-9612-895bc43c2683)
STN can also be extended to be 3D affine transformation.

**Comment by NorbertZheng** [comment]
## Reference
 - [2015 NIPS] [ST] [Spatial Transformer Networks](https://arxiv.org/abs/1506.02025).
